Tasks/DeployAzureResourceGroup/MachineGroupAgentExtensionManager.ts

```diff
@@ -0,0 +1,179 @@
+var computeManagementClient = require("azure-arm-compute");
+import util = require("util");
+import tl = require("vsts-task-lib/task");
+import azure_utils = require("./AzureUtil");
+import deployAzureRG = require("./DeployAzureRG");
+import Q = require("q");
+import constants = require("./Constants");
+
+export class MachineGroupAgentExtensionManager {
+    private taskParameters: deployAzureRG.AzureRGTaskParameters;
+    private credentials;
+    private subscriptionId: string;
+    private azureUtils;
+    private failureCount: number;
+    private successCount: number;
+    private errors: string;
+    private vmCount: number;
+    private deferred;
+    private operation: string;
+    private computeClient;
+
+    constructor(taskParameters: deployAzureRG.AzureRGTaskParameters) {
+        this.taskParameters = taskParameters;
+        this.credentials = this.taskParameters.credentials;
+        this.subscriptionId = this.taskParameters.subscriptionId;
+        this.azureUtils = new azure_utils.AzureUtil(this.taskParameters);
+        this.successCount = 0;
+        this.failureCount = 0;
+        this.vmCount = 0;
+        this.computeClient = new computeManagementClient(this.credentials, this.subscriptionId);
+        return this;
+    }
+
+    public async installMGExtension() {
+        try {
+            this.deferred = Q.defer<string>();
+            var operation = "installation";
+            var listOfVms = await this.azureUtils.getVMDetails();
+            this.vmCount = listOfVms.length;
+            if (this.vmCount == 0) {
+                this.deferred.resolve("");
+            }
+            for (var i = 0; i < listOfVms.length; i++) {
+                var vmName = listOfVms[i]["name"];
+                var resourceGroupName = this.taskParameters.resourceGroupName;
+                var extensionParameters = this.formExtensionParameters(listOfVms[i], operation);
+                this.log(tl.loc("AddExtension", extensionParameters["extensionName"], vmName));
+                var extensionName = extensionParameters["extensionName"];
+                var parameters = extensionParameters["parameters"];
+                var callback = this.givePostOperationCallBack(extensionName, vmName, operation);
+                this.computeClient.virtualMachineExtensions.createOrUpdate(resourceGroupName, vmName, extensionName, parameters, callback);
+            }
+            return this.deferred.promise;
+        }
+        catch (exception) {
+            this.log(tl.loc("MGAgentOperationOnAllVMsFailed", exception.message));
+            tl.setResult(tl.TaskResult.Failed, tl.loc("MGAgentOperationOnAllVMsFailed", exception.message));
+            this.deferred.resolve(tl.loc("MGAgentOperationOnAllVMsFailed", exception.message));
+        }
+    }
+
+    public async deleteMGExtension() {
+        try {
+            var operation = "uninstallation";
+            this.deferred = Q.defer<string>();
```

Comment:
I don't link the idea of keeping deferred as a member variable. It ties all simultaneous operations in extManager together.